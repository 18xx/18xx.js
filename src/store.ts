import { GameState } from './reducers/game';

export interface Store {
  getState: (hash: string) => Promise<GameState>;
  setState: (hash: string, body: GameState) => Promise<boolean>;
}
