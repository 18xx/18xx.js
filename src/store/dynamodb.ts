import * as AWS from 'aws-sdk';
import { DynamoDB } from 'aws-sdk';

import { GameState } from '../reducers/game';

import { Store } from '../store';

AWS.config.update({ region: 'ca-central-1' });

export default class DynamoDBStore implements Store {
  private ddb: DynamoDB;
  private ddbClient: DynamoDB.DocumentClient;

  constructor(url: string) {
    this.ddb = new AWS.DynamoDB({
      endpoint: url
    });
    this.ddbClient = new AWS.DynamoDB.DocumentClient({
      service: this.ddb
    });
    this.createTables();
  }

  public createTables(): void {
    const tableData: any = {
      AttributeDefinitions: [
        {
          AttributeName: 'key',
          AttributeType: 'S'
        }
      ],
      KeySchema: [
        {
          AttributeName: 'key',
          KeyType: 'HASH'
        }
      ],
      ProvisionedThroughput: {
        ReadCapacityUnits: '5',
        WriteCapacityUnits: '5'
      },
      TableName: 'gameState'
    };
    this.ddb.createTable(tableData).promise().then(
      (result: any) => console.log(result)
    ).catch(
      (result: any) => {
        // FIXME: Don't use exceptions for flow control
        if (result.code !== 'ResourceInUseException') {
          console.error(result);
        }
      }
    );
  }

  public getState(hash: string): Promise<GameState> {
    return this.ddbClient.get({
      Key: {
        key: hash
      },
      TableName: 'gameState',
    }).promise().then(result => {
      if (!result.Item) {
        return Promise.reject(`Could not find state: ${hash}`);
      }
      return result.Item.body;
    });
  }

  public setState(hash: string, body: GameState): Promise<boolean> {
    return this.ddbClient.put({
      Item: {
        body,
        key: hash,
      },
      TableName: 'gameState',
    }).promise().then(() => true);
  }
}
