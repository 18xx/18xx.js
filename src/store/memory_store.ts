import { Map } from 'immutable';

import { GameState } from '../reducers/game';

import { Store } from '../store';

export default class MemoryStore implements Store {
  private map: Map<string, GameState>;

  constructor() {
    this.map = Map<string, GameState>();
  }

  public getState(hash: string): Promise<GameState> {
    return Promise.resolve(this.map.get(hash));
  }

  public setState(hash: string, body: GameState): Promise<boolean> {
    this.map = this.map.set(hash, body);
    return Promise.resolve(true);
  }
}
