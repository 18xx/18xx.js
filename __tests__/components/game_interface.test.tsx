import { Map } from 'immutable';
import * as React from 'react';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';

import GameInterface, {
  GameInterfaceProps
} from '../../src/components/game_interface';

import { mapDefinition } from '../support/map_definition';
import { stubStore } from '../support/store';

describe('Game', () => {
  describe('#render', () => {
    const props: GameInterfaceProps = {
      gameName: '18xx',
      mapDef: mapDefinition,
      store: stubStore,
      tileFilter: Map(),
      tiles: Map(),
      tokens: Map(),
    };
    it('renders the game', () => {
      const subject: React.ReactElement<GameInterface> = (
        <Provider store={stubStore}>
          <GameInterface {...props} store={stubStore} />
        </Provider>
      );

      expect(renderer.create(subject)).toMatchSnapshot();
    });

    describe('open menus', () => {
      it('shows the tile menu', () => {
        const subject: React.ReactElement<GameInterface> = (
          <Provider store={stubStore}>
            <GameInterface {...props} store={stubStore} openMenu='TILE' />
          </Provider>
        );

        expect(renderer.create(subject)).toMatchSnapshot();
      });

      it('shows the token menu', () => {
        const subject: React.ReactElement<GameInterface> = (
          <Provider store={stubStore}>
            <GameInterface {...props} store={stubStore} openMenu='TOKEN' />
          </Provider>
        );

        expect(renderer.create(subject)).toMatchSnapshot();
      });

      it('shows the token context menu', () => {
        const subject: React.ReactElement<GameInterface> = (
          <Provider store={stubStore}>
            <GameInterface {...props} store={stubStore}
              openMenu='TOKEN_CONTEXT' />
          </Provider>
        );

        expect(renderer.create(subject)).toMatchSnapshot();
      });
    });
  });
});
