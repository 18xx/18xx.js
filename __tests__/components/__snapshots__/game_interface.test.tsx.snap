// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game #render open menus shows the tile menu 1`] = `
<div>
  <div
    id="tileMenu"
  />
  <div
    className="row"
  >
    <div
      className="col-sm-3 col-md-2 sidebar"
    >
      <h4>
        Recent Moves
      </h4>
      <ol
        reversed={true}
      />
    </div>
    <div
      className="col-sm-9 col-sm-offset-3 col-md-10 col-md-offset-2 main"
    >
      <svg
        height={224}
        id="mapBoard"
        width={387.9793808954285}
        xmlns="http://www.w3.org/2000/svg"
      >
        <svg
          className="hex"
          onClick={[Function]}
          x={0}
          y={0}
        >
          <polygon
            fill="red"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <text
            fontSize={16}
            textAnchor="middle"
            x={55.42562584220407}
            y={80}
          >
            One Town
          </text>
          <g>
            <polygon
              fill="black"
              points="110.85125168440814,60 110.85125168440814,68 96.99484522385713,64"
            />
            <polygon
              fill="black"
              points="86.60254037844386,110 79.67433714816835,114 76.2102355330306,100"
            />
          </g>
          <g>
            <g>
              <rect
                fill="brown"
                height={20}
                stroke="black"
                width={20}
                x={15.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={25.42562584220407}
                y={54}
              >
                50
              </text>
            </g>
            <g>
              <rect
                fill="gray"
                height={20}
                stroke="black"
                width={20}
                x={35.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={45.42562584220407}
                y={54}
              >
                60
              </text>
            </g>
            <g>
              <rect
                fill="green"
                height={20}
                stroke="black"
                width={20}
                x={55.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={65.42562584220407}
                y={54}
              >
                30
              </text>
            </g>
            <g>
              <rect
                fill="yellow"
                height={20}
                stroke="black"
                width={20}
                x={75.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={85.42562584220407}
                y={54}
              >
                20
              </text>
            </g>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={110.85125168440814}
          y={0}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <text
            fontSize={16}
            textAnchor="middle"
            x={55.42562584220407}
            y={100}
          >
            Threesville
          </text>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={55.42562584220407}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <g>
            <rect
              fill="#aaf"
              height={20}
              stroke="#333"
              width={20}
              x={45.42562584220407}
              y={14}
            />
            <text
              fontSize="13"
              textAnchor="middle"
              x={55.42562584220407}
              y={30}
            >
              40
            </text>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={166.27687752661222}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <g>
            <polygon
              fill="#f86"
              points="55.42562584220407,9 75.42562584220407,39 35.42562584220407,39"
              stroke="#333"
            />
            <text
              fontSize="13"
              textAnchor="middle"
              x={55.42562584220407}
              y={35}
            >
              120
            </text>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={277.12812921102034}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={387.9793808954285}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <svg
            height={128}
            viewBox="0 0 110.85125168440814 128"
            width={110.85125168440814}
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              fill="#bbb"
              points="0,32 55.42562584220407,0 110.85125168440814,32 110.85125168440814,96 55.42562584220407,128 0,96"
            />
            <g>
              <line
                stroke="black"
                strokeDasharray="1,0"
                strokeWidth={8}
                x1={0}
                x2={55.42562584220407}
                y1={63.99999999999999}
                y2={64}
              />
              <circle
                cx={55.42562584220407}
                cy={64}
                r="4"
              />
            </g>
            <svg
              onContextMenu={[Function]}
              x={35.42562584220407}
              y={44}
            >
              <circle
                className="city-circle"
                cx={20}
                cy={20}
                fill="white"
                r={19}
                stroke="#000"
                strokeWidth={2}
              />
            </svg>
            <text
              fill="black"
              fontSize={16}
              fontWeight="bold"
              textAnchor="end"
              x={108.85125168440814}
              y={46}
            >
              Preplaced
            </text>
          </svg>
        </svg>
      </svg>
    </div>
  </div>
</div>
`;

exports[`Game #render open menus shows the token context menu 1`] = `
<div>
  <div
    id="tokenMenu"
  >
    <button
      onClick={[Function]}
    >
      Remove Token
    </button>
  </div>
  <div
    className="row"
  >
    <div
      className="col-sm-3 col-md-2 sidebar"
    >
      <h4>
        Recent Moves
      </h4>
      <ol
        reversed={true}
      />
    </div>
    <div
      className="col-sm-9 col-sm-offset-3 col-md-10 col-md-offset-2 main"
    >
      <svg
        height={224}
        id="mapBoard"
        width={387.9793808954285}
        xmlns="http://www.w3.org/2000/svg"
      >
        <svg
          className="hex"
          onClick={[Function]}
          x={0}
          y={0}
        >
          <polygon
            fill="red"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <text
            fontSize={16}
            textAnchor="middle"
            x={55.42562584220407}
            y={80}
          >
            One Town
          </text>
          <g>
            <polygon
              fill="black"
              points="110.85125168440814,60 110.85125168440814,68 96.99484522385713,64"
            />
            <polygon
              fill="black"
              points="86.60254037844386,110 79.67433714816835,114 76.2102355330306,100"
            />
          </g>
          <g>
            <g>
              <rect
                fill="brown"
                height={20}
                stroke="black"
                width={20}
                x={15.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={25.42562584220407}
                y={54}
              >
                50
              </text>
            </g>
            <g>
              <rect
                fill="gray"
                height={20}
                stroke="black"
                width={20}
                x={35.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={45.42562584220407}
                y={54}
              >
                60
              </text>
            </g>
            <g>
              <rect
                fill="green"
                height={20}
                stroke="black"
                width={20}
                x={55.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={65.42562584220407}
                y={54}
              >
                30
              </text>
            </g>
            <g>
              <rect
                fill="yellow"
                height={20}
                stroke="black"
                width={20}
                x={75.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={85.42562584220407}
                y={54}
              >
                20
              </text>
            </g>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={110.85125168440814}
          y={0}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <text
            fontSize={16}
            textAnchor="middle"
            x={55.42562584220407}
            y={100}
          >
            Threesville
          </text>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={55.42562584220407}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <g>
            <rect
              fill="#aaf"
              height={20}
              stroke="#333"
              width={20}
              x={45.42562584220407}
              y={14}
            />
            <text
              fontSize="13"
              textAnchor="middle"
              x={55.42562584220407}
              y={30}
            >
              40
            </text>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={166.27687752661222}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <g>
            <polygon
              fill="#f86"
              points="55.42562584220407,9 75.42562584220407,39 35.42562584220407,39"
              stroke="#333"
            />
            <text
              fontSize="13"
              textAnchor="middle"
              x={55.42562584220407}
              y={35}
            >
              120
            </text>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={277.12812921102034}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={387.9793808954285}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <svg
            height={128}
            viewBox="0 0 110.85125168440814 128"
            width={110.85125168440814}
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              fill="#bbb"
              points="0,32 55.42562584220407,0 110.85125168440814,32 110.85125168440814,96 55.42562584220407,128 0,96"
            />
            <g>
              <line
                stroke="black"
                strokeDasharray="1,0"
                strokeWidth={8}
                x1={0}
                x2={55.42562584220407}
                y1={63.99999999999999}
                y2={64}
              />
              <circle
                cx={55.42562584220407}
                cy={64}
                r="4"
              />
            </g>
            <svg
              onContextMenu={[Function]}
              x={35.42562584220407}
              y={44}
            >
              <circle
                className="city-circle"
                cx={20}
                cy={20}
                fill="white"
                r={19}
                stroke="#000"
                strokeWidth={2}
              />
            </svg>
            <text
              fill="black"
              fontSize={16}
              fontWeight="bold"
              textAnchor="end"
              x={108.85125168440814}
              y={46}
            >
              Preplaced
            </text>
          </svg>
        </svg>
      </svg>
    </div>
  </div>
</div>
`;

exports[`Game #render open menus shows the token menu 1`] = `
<div>
  <div
    id="tokenMenu"
  >
    <span
      onClick={[Function]}
    >
      <svg
        className="token"
        height="40"
        onContextMenu={[Function]}
        opacity={1}
        viewBox="0 0 40 40"
        width="40"
      >
        <circle
          cx={20}
          cy={20}
          fill="#070A3F"
          r={16}
        />
        <rect
          fill="white"
          height={12}
          rx={1.75}
          ry={12}
          width={32}
          x={4}
          y={14}
        />
        <text
          fill="black"
          fontSize={10}
          fontWeight="normal"
          stroke="none"
          textAnchor="middle"
          x={20}
          y={24}
        >
          B&O
        </text>
      </svg>
    </span>
  </div>
  <div
    className="row"
  >
    <div
      className="col-sm-3 col-md-2 sidebar"
    >
      <h4>
        Recent Moves
      </h4>
      <ol
        reversed={true}
      />
    </div>
    <div
      className="col-sm-9 col-sm-offset-3 col-md-10 col-md-offset-2 main"
    >
      <svg
        height={224}
        id="mapBoard"
        width={387.9793808954285}
        xmlns="http://www.w3.org/2000/svg"
      >
        <svg
          className="hex"
          onClick={[Function]}
          x={0}
          y={0}
        >
          <polygon
            fill="red"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <text
            fontSize={16}
            textAnchor="middle"
            x={55.42562584220407}
            y={80}
          >
            One Town
          </text>
          <g>
            <polygon
              fill="black"
              points="110.85125168440814,60 110.85125168440814,68 96.99484522385713,64"
            />
            <polygon
              fill="black"
              points="86.60254037844386,110 79.67433714816835,114 76.2102355330306,100"
            />
          </g>
          <g>
            <g>
              <rect
                fill="brown"
                height={20}
                stroke="black"
                width={20}
                x={15.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={25.42562584220407}
                y={54}
              >
                50
              </text>
            </g>
            <g>
              <rect
                fill="gray"
                height={20}
                stroke="black"
                width={20}
                x={35.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={45.42562584220407}
                y={54}
              >
                60
              </text>
            </g>
            <g>
              <rect
                fill="green"
                height={20}
                stroke="black"
                width={20}
                x={55.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={65.42562584220407}
                y={54}
              >
                30
              </text>
            </g>
            <g>
              <rect
                fill="yellow"
                height={20}
                stroke="black"
                width={20}
                x={75.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={85.42562584220407}
                y={54}
              >
                20
              </text>
            </g>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={110.85125168440814}
          y={0}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <text
            fontSize={16}
            textAnchor="middle"
            x={55.42562584220407}
            y={100}
          >
            Threesville
          </text>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={55.42562584220407}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <g>
            <rect
              fill="#aaf"
              height={20}
              stroke="#333"
              width={20}
              x={45.42562584220407}
              y={14}
            />
            <text
              fontSize="13"
              textAnchor="middle"
              x={55.42562584220407}
              y={30}
            >
              40
            </text>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={166.27687752661222}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <g>
            <polygon
              fill="#f86"
              points="55.42562584220407,9 75.42562584220407,39 35.42562584220407,39"
              stroke="#333"
            />
            <text
              fontSize="13"
              textAnchor="middle"
              x={55.42562584220407}
              y={35}
            >
              120
            </text>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={277.12812921102034}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={387.9793808954285}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <svg
            height={128}
            viewBox="0 0 110.85125168440814 128"
            width={110.85125168440814}
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              fill="#bbb"
              points="0,32 55.42562584220407,0 110.85125168440814,32 110.85125168440814,96 55.42562584220407,128 0,96"
            />
            <g>
              <line
                stroke="black"
                strokeDasharray="1,0"
                strokeWidth={8}
                x1={0}
                x2={55.42562584220407}
                y1={63.99999999999999}
                y2={64}
              />
              <circle
                cx={55.42562584220407}
                cy={64}
                r="4"
              />
            </g>
            <svg
              onContextMenu={[Function]}
              x={35.42562584220407}
              y={44}
            >
              <circle
                className="city-circle"
                cx={20}
                cy={20}
                fill="white"
                r={19}
                stroke="#000"
                strokeWidth={2}
              />
            </svg>
            <text
              fill="black"
              fontSize={16}
              fontWeight="bold"
              textAnchor="end"
              x={108.85125168440814}
              y={46}
            >
              Preplaced
            </text>
          </svg>
        </svg>
      </svg>
    </div>
  </div>
</div>
`;

exports[`Game #render renders the game 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-sm-3 col-md-2 sidebar"
    >
      <h4>
        Recent Moves
      </h4>
      <ol
        reversed={true}
      />
    </div>
    <div
      className="col-sm-9 col-sm-offset-3 col-md-10 col-md-offset-2 main"
    >
      <svg
        height={224}
        id="mapBoard"
        width={387.9793808954285}
        xmlns="http://www.w3.org/2000/svg"
      >
        <svg
          className="hex"
          onClick={[Function]}
          x={0}
          y={0}
        >
          <polygon
            fill="red"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <text
            fontSize={16}
            textAnchor="middle"
            x={55.42562584220407}
            y={80}
          >
            One Town
          </text>
          <g>
            <polygon
              fill="black"
              points="110.85125168440814,60 110.85125168440814,68 96.99484522385713,64"
            />
            <polygon
              fill="black"
              points="86.60254037844386,110 79.67433714816835,114 76.2102355330306,100"
            />
          </g>
          <g>
            <g>
              <rect
                fill="brown"
                height={20}
                stroke="black"
                width={20}
                x={15.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={25.42562584220407}
                y={54}
              >
                50
              </text>
            </g>
            <g>
              <rect
                fill="gray"
                height={20}
                stroke="black"
                width={20}
                x={35.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={45.42562584220407}
                y={54}
              >
                60
              </text>
            </g>
            <g>
              <rect
                fill="green"
                height={20}
                stroke="black"
                width={20}
                x={55.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={65.42562584220407}
                y={54}
              >
                30
              </text>
            </g>
            <g>
              <rect
                fill="yellow"
                height={20}
                stroke="black"
                width={20}
                x={75.42562584220407}
                y={39}
              />
              <text
                textAnchor="middle"
                x={85.42562584220407}
                y={54}
              >
                20
              </text>
            </g>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={110.85125168440814}
          y={0}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <text
            fontSize={16}
            textAnchor="middle"
            x={55.42562584220407}
            y={100}
          >
            Threesville
          </text>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={55.42562584220407}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <g>
            <rect
              fill="#aaf"
              height={20}
              stroke="#333"
              width={20}
              x={45.42562584220407}
              y={14}
            />
            <text
              fontSize="13"
              textAnchor="middle"
              x={55.42562584220407}
              y={30}
            >
              40
            </text>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={166.27687752661222}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <g>
            <polygon
              fill="#f86"
              points="55.42562584220407,9 75.42562584220407,39 35.42562584220407,39"
              stroke="#333"
            />
            <text
              fontSize="13"
              textAnchor="middle"
              x={55.42562584220407}
              y={35}
            >
              120
            </text>
          </g>
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={277.12812921102034}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
        </svg>
        <svg
          className="hex"
          onClick={[Function]}
          x={387.9793808954285}
          y={96}
        >
          <polygon
            fill="#efe"
            points="110.85125168440814,32 55.42562584220407,0 0,32 0,96 55.42562584220407,128 110.85125168440814,96"
            stroke="black"
          />
          <svg
            height={128}
            viewBox="0 0 110.85125168440814 128"
            width={110.85125168440814}
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              fill="#bbb"
              points="0,32 55.42562584220407,0 110.85125168440814,32 110.85125168440814,96 55.42562584220407,128 0,96"
            />
            <g>
              <line
                stroke="black"
                strokeDasharray="1,0"
                strokeWidth={8}
                x1={0}
                x2={55.42562584220407}
                y1={63.99999999999999}
                y2={64}
              />
              <circle
                cx={55.42562584220407}
                cy={64}
                r="4"
              />
            </g>
            <svg
              onContextMenu={[Function]}
              x={35.42562584220407}
              y={44}
            >
              <circle
                className="city-circle"
                cx={20}
                cy={20}
                fill="white"
                r={19}
                stroke="#000"
                strokeWidth={2}
              />
            </svg>
            <text
              fill="black"
              fontSize={16}
              fontWeight="bold"
              textAnchor="end"
              x={108.85125168440814}
              y={46}
            >
              Preplaced
            </text>
          </svg>
        </svg>
      </svg>
    </div>
  </div>
</div>
`;
